### PR TITLE
Make necessary changes to support asymmetric actor critic.

### PR DESCRIPTION
**Pull request by kevinzakka**

### CODE REVIEW

File: brax/training/agents/ppo/networks.py [hunk 1]
```diff
@@ -68,7 +68,9 @@ def make_ppo_networks(
     .identity_observation_preprocessor,
     policy_hidden_layer_sizes: Sequence[int] = (32,) * 4,
     value_hidden_layer_sizes: Sequence[int] = (256,) * 5,
-    activation: networks.ActivationFn = linen.swish) -> PPONetworks:
+    activation: networks.ActivationFn = linen.swish,
+    policy_state_key: str = 'state',
+    value_state_key: str = 'state') -> PPONetworks:
```
Comment (btaba) [inline brax/training/agents/ppo/networks.py:73]: What would you pass in as `value_state_key` if you wanted `state` and `state_privileged` in the value network?

File: brax/training/agents/ppo/train.py [hunk 2]
```diff
@@ -336,7 +335,7 @@ def f(carry, unused_t):
     # Update normalization params and normalize observations.
     normalizer_params = running_statistics.update(
         training_state.normalizer_params,
-        data.observation if ndarray_obs else data.observation['state'],
+        data.observation,
```
Comment (btaba) [inline brax/training/agents/ppo/train.py:338]: Can you add a test in train_test.py that actually makes use of asymmetric obs? It's hard to tell why you chose this impl without seeing it used in practice

**Review by btaba — CHANGES_REQUESTED** [review]

File: brax/training/agents/ppo/networks.py (hunk 1 above)
Comment (kevinzakka) [reply brax/training/agents/ppo/networks.py:73]: In my current implementation, the env defines privileged state as the union of state + additional privileged information.

File: brax/training/agents/ppo/train.py (hunk 2 above)
Comment (btaba) [reply brax/training/agents/ppo/train.py:338]: Please fix

Comment (kevinzakka) [reply brax/training/agents/ppo/train.py:338]: Added!

File: brax/training/agents/ppo/train_test.py [hunk 3]
```diff
@@ -131,6 +133,53 @@ def get_offset(rng):
...
+      })
+
+    class FastAsymmetric(Fast):
+      def reset(self, rng):
+        reset_state = super().reset(rng)
+        return _add_privileged_state(reset_state, self.observation_size)
+
+      def step(self, state, action):
+        step_state = super().step(state, action)
+        return _add_privileged_state(step_state, self.observation_size)
```
Comment (btaba) [inline brax/training/agents/ppo/train_test.py:156]: can you just modify the Fast env with a flag?

File: brax/training/agents/ppo/train_test.py [hunk 4]
```diff
@@ -131,6 +133,53 @@ def get_offset(rng):
...
+
+    network_factory = functools.partial(
+      ppo_networks.make_ppo_networks,
+      policy_state_key='state',
+      value_state_key='privileged_state'
+    )
+
+    _, _, _ = ppo.train(
```
Comment (btaba) [inline brax/training/agents/ppo/train_test.py:164]: can you make an assert on the value/policy param shape, just to make sure the networks are doing what we expect?

File: brax/training/networks.py [hunk 5]
```diff
@@ -94,7 +94,8 @@ def make_policy_network(
     hidden_layer_sizes: Sequence[int] = (256, 256),
     activation: ActivationFn = linen.relu,
     kernel_init: Initializer = jax.nn.initializers.lecun_uniform(),
-    layer_norm: bool = False) -> FeedForwardNetwork:
+    layer_norm: bool = False,
+    state_key: str = 'state') -> FeedForwardNetwork:
```
Comment (btaba) [inline brax/training/networks.py:98]: maybe this should be `obs_key`? Either that or add a docstring, it's a fairly unintuitive field as opposed to the others

**Review by btaba — CHANGES_REQUESTED** [review]
Lookin' good, just a few nits

File: brax/training/agents/ppo/train_test.py (hunk 3 above)
Comment (kevinzakka) [reply brax/training/agents/ppo/train_test.py]: Done

File: brax/training/agents/ppo/train_test.py (hunk 4 above)
Comment (kevinzakka) [reply brax/training/agents/ppo/train_test.py:171]: Done

File: brax/envs/fast.py [hunk 6]
```diff
@@ -29,8 +29,18 @@ def __init__(self, **kwargs):
     self._reset_count = 0
     self._step_count = 0
     self._use_dict_obs = kwargs.get('use_dict_obs', False)
+    self._asymmetric_obs = kwargs.get('asymmetric_obs', False)
+    if self._asymmetric_obs and not self._use_dict_obs:
+      raise ValueError('asymmetric_obs requires use_dict_obs=True')
+
+  def _get_obs(self):
+    obs = {'state': jp.zeros(2)} if self._use_dict_obs else jp.zeros(2)
```
Comment (btaba) [inline brax/envs/fast.py:37]:
```
if not self._use_dict_obs:
  return jp.zeros(2)

obs = {'state': jp.zeros(2)}
if self._use_asymmetric_obs:
  obs['privilged_state'] = ...

return obs
```

File: brax/envs/fast.py [hunk 7]
```diff
@@ -70,6 +78,11 @@ def step_count(self):
 
   @property
   def observation_size(self):
+    if self._use_dict_obs:
```
Comment (btaba) [inline brax/envs/fast.py:81]: can use flattened logic like above

File: brax/training/agents/ppo/train_test.py (hunk 4 above)
Comment (btaba) [reply brax/training/agents/ppo/train_test.py:171]: I'm confused, there are no asserts. I figured you can just grab the params returned from `ppo.train` and assert those input shapes

**Review by btaba — CHANGES_REQUESTED** [review]

File: brax/envs/fast.py (hunk 6 above)
Comment (kevinzakka) [reply brax/envs/fast.py]: Done

File: brax/envs/fast.py (hunk 7 above)
Comment (kevinzakka) [reply brax/envs/fast.py]: Done

File: brax/training/agents/ppo/train_test.py (hunk 4 above)
Comment (kevinzakka) [reply brax/training/agents/ppo/train_test.py]: Done

**Review by btaba — APPROVED** [review]

**Comment by btaba** [comment]
sweet ty 🙌

**Pull request merged**